src/main/java/edu/cornell/kfs/sys/aws/batch/AwsSecretValidationStep.java

```diff
@@ -0,0 +1,167 @@
+            errorMessage = errorMessage + " The expected user name was not found.";
+        }
+
+        if (StringUtils.isNotBlank(errorMessage)) {
+            throw new RuntimeException(errorMessage);
+        } else {
+            LOG.info("confirmAmazonSecretValidationShared, the shared secrets contained the expected values");
+        }
+    }
+
+    private Map<String, Object> buildNewSecretValue() {
+        Map<String, Object> values = new HashMap<>();
+        values.put(ACCESS_TOKEN, UUID.randomUUID().toString());
+        values.put(REFRESH_TOKEN, UUID.randomUUID().toString());
+        values.put(ACCESS_TOKEN_EXPIRATION_DATE, Calendar.getInstance(Locale.US).getTime());
+        return values;
+    }
+    
+    protected void updateAmazonSecretValidationInstance(AmazonSecretValidationInstance instance,
+            Map<String, Object> newSecretValues) {
+        instance.setAccess_token(newSecretValues.get(ACCESS_TOKEN).toString());
+        instance.setRefresh_token(newSecretValues.get(REFRESH_TOKEN).toString());
+        instance.setAccess_token_expiration_date((Date) newSecretValues.get(ACCESS_TOKEN_EXPIRATION_DATE));
+    }
+
+    protected void confirmInstanceSecretValues(Map<String, Object> expectedSecretValues,
+            AmazonSecretValidationInstance instanceValues) {
+        String errorMessage = StringUtils.EMPTY;
+        if (!StringUtils.equals(instanceValues.getAccess_token(),
+                expectedSecretValues.get(ACCESS_TOKEN).toString())) {
+            errorMessage = errorMessage + "Did not update ACCESS token as expected.";
+        }
+        if (!StringUtils.equals(instanceValues.getRefresh_token(),
+                expectedSecretValues.get(REFRESH_TOKEN).toString())) {
+            errorMessage = errorMessage + " Did not update REFRESH token as expected.";
+        }
+        if (!instanceValues.getAccess_token_expiration_date()
+                .equals((Date) expectedSecretValues.get(ACCESS_TOKEN_EXPIRATION_DATE))) {
+            errorMessage = errorMessage + " Did not update access token EXPIRATION DATE as expected.";
+        }
+
+        if (StringUtils.isNotBlank(errorMessage)) {
+            throw new RuntimeException(errorMessage);
+        }
+    }
+    
+    protected void validateUsingNewAWSServiceFromSpring() {
+        AwsSecretService service = SpringContext.getBean(AwsSecretService.class);
+        try {
+            AmazonSecretValidationShared sharedSpaceSecrets = service.getPojoFromAwsSecret(AWS_SECRET_NAME_VALIDATION_SHARED, 
+                    false, AmazonSecretValidationShared.class);
+            confirmAmazonSecretValidationShared(sharedSpaceSecrets);
+            service.logCacheStatus();
+        } catch (JsonProcessingException e) {
+            LOG.error("validateUsingNewAWSServiceFromSpring, had an error calling AWS Secret Service ", e);
+            throw new RuntimeException(e);
+        }
+    }
+    
+    protected void waitAfterUpdatingSecretBeforeGettingSecret() {
```

Comment:
Is there caching in AWS itself or in the AWS Java libraries that makes it necessary to wait after performing an update? If such a wait is necessary, that could make it more difficult to get Concur to use the AWS Secrets Manager properly.

I see that AWS also has an official Java library to help with AWS secrets caching. Would that library address any of the wait-after-update concerns, or would it be simpler to just go with our custom caching solution?

If we expect this kind of waiting to be more commonplace with the move to AWS secrets, it might be helpful to make this method more accessible to other areas of the code.

Reply:
The cache library I found related to secrets manager API is https://github.com/aws/aws-secretsmanager-caching-java ..... but that only works with the 1.x version of the API.  Per recommendation from AWS, we are now using the latest 2.x version of the API.  Were you able to find a cache service with the new API?

In a ticket we opened with Amazon, I asked if there was latency that I should be aware of.   Amazon support responded with "The latency shouldn't be more than a couple of seconds if the secret was properly updated."

In our normal course of business we won't need to do this wait as we will be keeping the value in cache.  I found where if I updated a secret, cleared the cache and then immediately checked the secret value, I wasn't retrieving the the updated value.  Adding a few second wait seems to have resolved the problem.  In normal course of business, we will use cache, so I don't think this wait will be needed.  Should I change this code to clarify the purpose?  Do you think a wait will be needed in a normal business process?

Reply:
Thanks for the clarification on the caching library. I see now that it only supports the 1.x API like you had said. (It does have some pending PRs to add support for the 2.x API, but I don't know when they will finally move forward.) I'm fine with using our own custom caching in the meantime.

I can't think of any batch jobs yet where we would need to perform a manual wait. With the Concur Event Notification process, I think the read/write of the tokens is handled through the ConcurAccessTokenService, so an extra wait may not be needed as long as our processes all go through this service to read/write the credential data (and where this service would have its own unique AWS secret service instance).

As for the web page that handles manual access token read/write, I'm not sure if it would be worth adding a wait in that spot. The page will need to manually clear the cache prior to at least its initial load; however, as long as there aren't multiple users changing the token at once and as long as we're not using this page around the time of the Concur job, we may not need to add a wait for this page.

If you still plan to keep this wait in place for at least the sample AWS validation job, I would suggest renaming the method to make the intent clearer. The same goes for the related method in the integration test.

Reply:
I think we need the wait for these test cases, but am having a hard time coming up with a good name, how about waitFiveSecondsAfterUpdatingSecretAndClearingCacheBeforeRetrieveSecretFromAws?